### PR TITLE
Add ADR for block level sync

### PR DESCRIPTION
**Pull request by prabhanshuguptagit**
**Story card:** [ch3265](https://app.clubhouse.io/simpledotorg/story/3265/add-adr-for-block-level-sync-flexible-sync-radius)

Block level syncing will become mainline code in #2430. The only documentation around it is [this PRD](https://docs.google.com/document/d/1Cflct0Y-44IRUVw_5-NptcnNSX1UgAPBiXqoXHq22io/edit) which is a bit cluttered. 

This adds a retroactive ADR on Region level syncing and a small how-to on switching sync radius for the future.

### CODE REVIEW

File: doc/arch/014-region-level-sync.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# Region level sync
+September 2020 (accepted)
+
+_This ADR has been added retroactively in Apr 2021 to capture our switch to block-level syncing._
```
Comment (ssrihari) [inline doc/arch/014-region-level-sync.md:4]: This is essentially the status section. Perhaps we can move that up here and coalesce with this part.

File: doc/arch/014-region-level-sync.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+## Context
+
+Users in large districts reported that the Simple app was running very slow, making the app near-unusable.
+The slowdown was correlated to the volume of patient data synced to the user’s phone. We realised that the amount of data being stored on the device had to be reduced for better long-term performance.
```
Comment (ssrihari) [inline doc/arch/014-region-level-sync.md:9]: > correlated

Should we say it was _caused_ by it?

File: doc/arch/014-region-level-sync.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+Currently we sync the entire district's records to a user's phone. Some of the large districts have upto 50,000 patients, which can amount to 400-500 MB of data.
```
Comment (ssrihari) [inline doc/arch/014-region-level-sync.md:11]: Is there a target storage volume (or a range) on the device at which the app performs optimally? We should ideally have that.

File: doc/arch/014-region-level-sync.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+A district typically has between 1-20 blocks. From trends in IHCI, we found that patients rarely move between facilities across blocks. Patients that have a BP taken in more than 1 block is around 2% or under, with the exceptions of: Sindhudurg (9.8%), Hoshiarpur (5.3%), Bathinda (3.1%).
+This means that we can sync only a block's data to the user's phone and be reasonably confident about finding patients on the app.
```
Comment (ssrihari) [inline doc/arch/014-region-level-sync.md:14]: > we found that patients rarely move between facilities across blocks

We have an average of ~600 duplicates across blocks per month in the last few months. Is this still what we would consider rare? Perhaps we should consider rewording accordingly.

File: doc/arch/014-region-level-sync.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+Accepted. This feature was released to all users by Feb 2020.
+
+## Consequences
+
+- The Simple app might not be able to find patients who moved from one block to another.
+- Block is currently a freeform text field on the `Facility` model.
+  It needs to be a first-class entity to make block-level syncing possible.
+  This is introduced through the `Region` model.
+- When a patient gets registered across blocks as a duplicate, we will need to identify them and merge their data.
```
Comment (ssrihari) [inline doc/arch/014-region-level-sync.md:46]: Should we also add that we plan to implement online patient lookup to account for the cases in which local lookup fails?

File: doc/arch/014-region-level-sync.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
+# Region level sync
+September 2020 (accepted)
+
+_This ADR has been added retroactively in Apr 2021 to capture our switch to block-level syncing._
+
+## Context
```
Comment (ssrihari) [inline doc/arch/014-region-level-sync.md:6]: If there's a PRD, we should link to it.

File: doc/arch/014-region-level-sync.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  This is important in case we need to change the kind of records that are synced to the app in the future.
+  See [Block level sync manual](../wiki/adjusting-sync-boundaries.md) for how it works.
+
+### On the app
+- Users can continue selecting any facility in their district when switching facilities.
+- Users can continue selecting any facility in their district when scheduling a patient’s next visit or preferred facility. 
+- It is possible that a patient will visit a facility outside their block and their record will not be found on the user’s device. In this case the user should 
+    - Scan the patient’s BP passport if they have one.
+    - Register the patient again, as if they were new. Make sure to attach their existing BP passport to the registration.
+    - The duplicate patient records will be merged by the Simple team later.
```
Comment (ssrihari) [inline doc/arch/014-region-level-sync.md:34]: How was the move to block sync managed end to end?

1. How is the change in sync region detected on the app?
2. How do we remove the previous block's data on the app when the change happens?
3. How do we ensure that all data is synced before removal?
4. How did we plan for resource utilisation increase on the backend to accommodate full syncs?

This is important context that will be useful the next time we decide to change the sync region.

Comment (ssrihari) [reply doc/arch/014-region-level-sync.md:34]: Ah, I see most of this was captured in the wiki 😅 . Very cool. I think only 3 is unanswered. Would be good to go into some detail overall therein. If there's a related android ADR, would be super relevant to link to it.

**Review by ssrihari — COMMENTED** [review]
👌  glad you took the time to write this down @prabhanshuguptagit

**Review by harimohanraj89 — CHANGES_REQUESTED** [review]
My thoughts.

File: doc/arch/014-region-level-sync.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+## Status
+
+Accepted. This feature was released to all users by Feb 2020.
+
+## Consequences
+
+- The Simple app might not be able to find patients who moved from one block to another.
+- Block is currently a freeform text field on the `Facility` model.
+  It needs to be a first-class entity to make block-level syncing possible.
+  This is introduced through the `Region` model.
```
[the remark this answers is not in the document]
Comment (prabhanshuguptagit) [reply doc/arch/014-region-level-sync.md:45]: Yes

File: doc/wiki/adjusting-sync-boundaries.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+A sync region can be any region in the `Region` hierarchy (state, district, block).
+
+This is a guide on how region level sync works and how to transition from one using one sync region to another sync region.
+
+### At the heart of it
+- The server sets the `sync_region_id` for each facility in the facility payload.
+- The app picks the current facility's `sync_region_id` as the sync region.
+- The app requests for the sync region's records to be synced by sending the `sync_region_id` in the `X-SYNC-REGION-ID` header.
+- The server returns data for the requested sync region's `syncable_patients`.
+  The criteria for selecting these patients is currently defined [in the Region model](https://github.com/simpledotorg/simple-server/blob/5a6416760e016a2ba2d12f46e46135d11927497c/app/models/region.rb#L130).
```
[the remark this answers is not in the document]
Comment (prabhanshuguptagit) [reply doc/wiki/adjusting-sync-boundaries.md:14]: This is the only place, other than the PRD.

**Review by harimohanraj89 — APPROVED** [review]
My feedback is addressed 🚀

**Pull request merged**